design-proposals/instancetype.kubevirt.io/expand-during-submission.md

```diff
@@ -0,0 +1,301 @@
+configurable was enabled in an environment with a large number of existing
+`VirtualMachines` using instance types and preferences.
+
+### Immutable instance types and preferences
+
+The need to retain point in time revisions of instance types and preferences is
+due to the simple fact that in the current implementation these resources are
+mutable and can change over time. Thus to ensure we always get the same
+`VirtualMachineInstance` at runtime revisions need to be taken and referenced
+from the `VirtualMachine`.
+
+We could possibly remove this requirement by making these object immutable and
+thus dropping the need capture and reference `ControllerRevisions` from the
+`VirtualMachines` at all.
+
+This however still retains the need for additional logic in more complex
+`VirtualMachine` lifecycle operations where we need to expand these now
+immutable resources in the `VirtualMachine`.
+
+### Expand by default and deprecate revision references
+
+We could alter the default `policy` to `expand` and in doing so deprecate the
+revision `reference` behaviour for eventual removal from the project ahead of
+`instancetype.kubevirt.io` finally making it to `v1`.
+
+### Keep existing behaviour
+
+Ultimately we can also decide not to implement the core proposal or any of the
+above alternatives and continue to support the original revision based flows.
+Supporting users and third-party integrators with better documentation and
+tooling for VirtualMachines referencing instance types or preferences.
+
+## API Examples
+
+The default `policy` will be `reference`.
+
+A cluster admin can default the policy of all new `VirtualMachines` by setting
+`expand` within the `KubeVirt` `CR`:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  name: kv
+spec:
+  configuration:
+    instancetype:
+      policy: expand
+    preference:
+      policy: expand
+```
+
+Additionally a VM owner can explicitly request `expand` by setting the policy
+directly on the `VirtualMachine`:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  instancetype:
+    name: foo
+    policy: expand
+  preference:
+    name: bar
+    policy: expand
+```
+
+Likewise users can explicitly request the `reference` policy to counter a
+different default being defined in the `KubeVirt` `CR`:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  instancetype:
+    name: foo
+    policy: reference
+```
+
+Users can also expand an existing `VirtualMachine` by removing any referenced
+revision from a matcher to invoke the expand behaviour, for example starting
+with the following:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  instancetype:
+    name: foo
+    revisionName: bar
+```
+
+A user can explicitly ask for the associated instance type to be expanded into
+the `VirtualMachine` by removing `revisionName` and setting `policy` to
+`expand`:
```

Comment:
Isn't adding `policy:expand` explicit enough? Why removing `revisionName` is required?

Reply:
The current proposal stops short of expanding all VirtualMachines and would use `revisionName` to determine if the VirtualMachine had already been seen by the controller and revisions stored. If we went with the alternative of expanding all regardless of a revision already being captured then `policy` being set to `expand` would be enough to trigger the behaviour.